### PR TITLE
Add docs about middlewares

### LINKED ISSUE #950: Documentation About Middleware
Add documentation about what middleware does and concrete examples

- [ ] Do a pass or middleware name to be more explicit

### PR DESCRIPTION
**Pull request by sundowndev**
| Q                 | A
| ----------------- | ---
| 🐛 Bug fix?       | no
| 🚀 New feature?   | no
| ⚠ Deprecations?   | no
| ❌ BC Break       | no
| 🔗 Related issues | Fix #950
| ❓ Documentation  | no 

## Description

This PR adds docs about middlewares: what middlewares does and how to create one.

### CODE REVIEW

**Review by eliecharra — DISMISSED** [review]
I'm good with the content, maybe @wbeuil can do a wording review ?

File: docs/middlewares.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Middlewares
+
+Main goal of middleware is to make IaC and remote resources set to match. For this we filter some default resources, mutate or remove some fields in resources or even create and delete new resources.
```
Comment (wbeuil) [inline docs/middlewares.md:3]: `The main goal of middlewareS is to reconciliate IaC and remote resources. For this we filter default resources, mutate, or remove fields in resources ...`

File: docs/middlewares.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+	// ...
+
+	*remoteResources = newRemoteResources
+  
+  return nil
+}
+```
+
+In the above example, we define a middleware called `AwsDefaultRoute` that will modify remote resources. Middleware can access two `*[]*resource.Resource` : the one from IaC side and the other one from remote side and the goal is to rework theses slice to remove false positive drifts. Notice both `remoteResources` and `resourcesFromState` variables are pointers, which mean middlewares can perform mutations on resources before the comparison is made.
```
Comment (wbeuil) [inline docs/middlewares.md:24]: `two *[]*resource.Resource: IaC resources first and then remote resources. The goal ...` (no space before : )
`these slices to ...`

File: docs/middlewares.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+## Different kind of middlewares
+
+1) Help driftctl match IaC resources and remote resources.
+2) Filter noise from provider default resources
```
Comment (wbeuil) [inline docs/middlewares.md:29]: `Filter noiseS`

File: docs/middlewares.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+3) Resource transformation
+4) Specific edge cases
+
+## Examples
+
+1) `aws_route_table_expander` explode inline route in dedicated resources
+2) `aws_default_route` ignore route created by default when creating a table **if they are not managed in IaC**
+3) `aws_iam_user_policy_attachment` and `aws_iam_role_policy_attachment` transformed to `aws_iam_policy_attachment`
+4) `route53_records_id_reconcilier` that rework ID's to match terraform ones
```
Comment (wbeuil) [inline docs/middlewares.md:38]: `Terraform ones`

**Review by wbeuil — CHANGES_REQUESTED** [review]

**Review by wbeuil — APPROVED** [review]

**Pull request merged**